include/pisa/sequence/partitioned_sequence.hpp

```diff
@@ -1,427 +1,412 @@
-                     m_params);
+            if (ub_it.first == m_upper_bounds.size()) {
+                return move(size());
             }
 
-            global_parameters m_params;
-            uint64_t m_partitions;
-            uint64_t m_endpoints_offset;
-            uint64_t m_endpoint_bits;
-            uint64_t m_sequences_offset;
-            uint64_t m_size;
-            uint64_t m_universe;
-
-            uint64_t m_position;
-            uint64_t m_cur_partition;
-            uint64_t m_cur_begin;
-            uint64_t m_cur_end;
-            uint64_t m_cur_base;
-            uint64_t m_cur_upper_bound;
-
-            bit_vector const* m_bv;
-            compact_elias_fano::enumerator m_sizes;
-            compact_elias_fano::enumerator m_upper_bounds;
-            base_sequence_enumerator m_partition_enum;
+            switch_partition(ub_it.first - 1);
+            return next_geq(lower_bound);
+        }
+
+        void switch_partition(uint64_t partition)
+        {
+            assert(m_partitions > 1);
+
+            uint64_t endpoint = partition
+                ? (m_bv->get_word56(m_endpoints_offset + (partition - 1) * m_endpoint_bits)
+                   & ((uint64_t(1) << m_endpoint_bits) - 1))
+                : 0;
+
+            uint64_t partition_begin = m_sequences_offset + endpoint;
+            m_bv->data().prefetch(partition_begin / 64);
+
+            m_cur_partition = partition;
+            auto size_it = m_sizes.move(partition);
+            m_cur_end = size_it.second;
+            m_cur_begin = m_sizes.prev_value();
+
+            auto ub_it = m_upper_bounds.move(partition + 1);
+            m_cur_upper_bound = ub_it.second;
+            m_cur_base = m_upper_bounds.prev_value() + (partition ? 1 : 0);
+
+            m_partition_enum = base_sequence_enumerator(
+                *m_bv,
+                partition_begin,
+                m_cur_upper_bound - m_cur_base + 1,
+                m_cur_end - m_cur_begin,
+                m_params);
+        }
+
+        global_parameters m_params;
+        uint64_t m_partitions;
+        uint64_t m_endpoints_offset;
+        uint64_t m_endpoint_bits;
+        uint64_t m_sequences_offset;
+        uint64_t m_size;
+        uint64_t m_universe;
+
+        uint64_t m_position;
+        uint64_t m_cur_partition;
+        uint64_t m_cur_begin;
+        uint64_t m_cur_end;
+        uint64_t m_cur_base;
+        uint64_t m_cur_upper_bound;
+
+        bit_vector const* m_bv;
+        compact_elias_fano::enumerator m_sizes;
+        compact_elias_fano::enumerator m_upper_bounds;
+        base_sequence_enumerator m_partition_enum;
+    };
+
+  private:
+    template <typename Iterator>
+    static std::vector<uint32_t> compute_partition(
+        Iterator begin,
+        uint64_t universe,
+        uint64_t n,
+        global_parameters const& params,
+        uint64_t fix_cost = 64,
+        double eps1 = 0.03,
+        double eps2 = 0.3,
+        double eps3 = 0.01)
```

Comment:
Can we add some comments as to what these four are? Just a few words and maybe a reference to some other resource/paper if that's more difficult to explain.